### PR TITLE
Add segwit signatures

### LINKED ISSUE #6: Add sign/verify for SegWit P2SH addresses
> For TREZOR we implemented a way how to Sign and Verify SegWit P2SH addresses. This 
> is achieved by introducing extra `v` values used during signing.
> 
> Currently the following are used:
> 
> * 27-30 uncompressed pubkey (base58)
> * 31-34 compressed pubkey (base58)
> 
> We added the following:
> 
> * 35-38 p2sh segwit pubkey (base58)
> * 39-42 segwit pubkey (bech32)
> 
> [Unit tests for reference](https://github.com/trezor/python-trezor/blob/master/tests/device_tests/test_msg_signmessage_segwit.py)

Reference https://github.com/bitcoinjs/bitcoinjs-lib/issues/880

### PR DESCRIPTION
**Pull request by junderw**
Fixes #6 

Needs tests.

### CODE REVIEW

**Comment by dabura667** [comment]
Needs to check for invalid segwitType and throw.

**Comment by junderw** [comment]
> Needs to check for invalid segwitType and throw.

Added check.

**Comment by junderw** [comment]
Added tests. Should be good for review @prusnak @dcousens @fanatid

**Comment by prusnak** [comment]
I think `'base58'` and `'bech32'` are weird choices for segwitType. I would expect `p2wpkh-in-p2sh` and `p2wpkh` or something similar.

**Comment by junderw** [comment]
@prusnak Fixed.

File: index.js [hunk 1]
```diff
@@ -1,9 +1,15 @@
 var bs58check = require('bs58check')
+var bech32 = require('bech32')
 var bufferEquals = require('buffer-equals')
 var createHash = require('create-hash')
 var secp256k1 = require('secp256k1')
 var varuint = require('varuint-bitcoin')
 
+var SEGWIT_TYPES = {
+  P2WPKH: 'p2wpkh',
+  P2SH_P2WPKH: 'p2wpkh-in-p2sh'
```
[the remark this answers is not in the document]
Comment (junderw) [reply index.js]: Sure why not

**Comment by junderw** [comment]
@dcousens switched to p2sh(p2wpkh)

File: index.js [hunk 2]
```diff
@@ -1,9 +1,15 @@
...
+var bech32 = require('bech32')
 var bufferEquals = require('buffer-equals')
 var createHash = require('create-hash')
 var secp256k1 = require('secp256k1')
 var varuint = require('varuint-bitcoin')
 
+var SEGWIT_TYPES = {
+  P2WPKH: 'p2wpkh',
+  P2SH_P2WPKH: 'p2sh(p2wpkh)'
+}
```
Comment (dcousens) [inline index.js]: IMHO,  you should additionally support `P2PKH`,  as that is the default behaviour.

Comment (junderw) [reply index.js]: I was trying not to break backwards compatibility... ie. passing nothing for segwitType will == exactly the same as before.

I could remove the compressed boolean arg and replace it with a type arg that would include segwit types AND p2pkh and p2pkh-uncompressed

**Comment by zooooooooo** [comment]
Any word on status of this?

**Comment by junderw** [comment]
bump

I think this is good to merge.

**Comment by EvilJordan** [comment]
I just wanted to say thanks for doing this work, and put out the good energy that everyone adopts it as the standard implementation coughtrezorcough.

**Pull request merged**